lib/commands/audit.js

```diff
@@ -1,8 +1,368 @@
+    }
+
+    const invalid = Array.from(this.invalid).sort(sortAlphabetically)
+    const missing = Array.from(this.missing).sort(sortAlphabetically)
+
+    const verified = invalid.length === 0 && missing.length === 0
+
+    if (!verified) {
+      this.exitCode = 1
+    }
+
+    const end = process.hrtime.bigint()
+    const elapsed = end - start
+
+    if (this.npm.config.get('json')) {
+      this.appendOutput(this.makeJSON({ invalid, missing }))
+    } else {
+      const auditedPlural = this.audited > 1 ? 's' : ''
+      const timing = `audited ${this.audited} package${auditedPlural} in ` +
+                     `${Math.floor(Number(elapsed) / 1e9)}s`
+      const verifiedPrefix = verified ? 'verified registry signatures, ' : ''
+      this.appendOutput(`${verifiedPrefix}${timing}\n`)
+
+      if (this.verified && !verified) {
+        const verifiedClr = this.npm.color ? chalk.bold('verified') : 'verified'
+        const msg = this.verified === 1 ?
+          `${this.verified} package has a ${verifiedClr} registry signature\n` :
+          `${this.verified} packages have ${verifiedClr} registry signatures\n`
+        this.appendOutput(msg)
+      }
+
+      if (missing.length) {
+        const logMissing = this.npm.config.get('log-missing-names')
+        const missingClr = this.npm.color ? chalk.bold(chalk.magenta('missing')) : 'missing'
+        const msg = missing.length === 1 ?
+          `package has a ${missingClr} registry signature` :
+          `packages have ${missingClr} registry signatures`
+        this.appendOutput(
+          `${missing.length} ${msg} but the registry is ` +
+          `providing signing keys${logMissing ? ':\n' : ''}`
+        )
+        if (logMissing) {
+          this.appendOutput(this.humanOutput(missing))
+        } else {
+          this.appendOutput(`  run \`npm audit signatures --log-missing-names\` for details`)
+        }
+      }
+
+      if (invalid.length) {
+        const invalidClr = this.npm.color ? chalk.bold(chalk.red('invalid')) : 'invalid'
+        const msg = invalid.length === 1 ?
+          `${invalid.length} package has an ${invalidClr} registry signature:\n` :
+          `${invalid.length} packages have ${invalidClr} registry signatures:\n`
+        this.appendOutput(
+          `${missing.length ? '\n' : ''}${msg}`
+        )
+        this.appendOutput(this.humanOutput(invalid))
+        const tamperMsg = invalid.length === 1 ?
+          `\nSomeone might have tampered with this package since it was ` +
+          `published on the registry!\n` :
+          `\nSomeone might have tampered with these packages since they where ` +
+          `published on the registry!\n`
+        this.appendOutput(tamperMsg)
+      }
+    }
+  }
+
+  findAllRegistryUrls (edges) {
+    return new Set(Array.from(edges, (edge) => {
+      const spec = this.getEdgeSpec(edge)
+      if (!spec) {
+        return
+      }
+      return this.getSpecRegistry(spec)
```

Comment:
I think we might want to try and get the registry url from the package's resolved tarball url to prevent a mirror/proxy from easily circumventing signature verification but omitting to route the keys endpoint. They would then need to modify the packument content and proxy tarball requests to get around this which is a lot more visible as it shows up in the lockfile.

@wraithgar thoughts on this? Could we get the resolved tarball url without fetching the manifest here again? Can we get it from the lockfile?